src/contracts/examples/redstone-oev/RedstoneAdapterAtlasWrapper.sol

```diff
@@ -88,64 +100,87 @@ contract RedstoneAdapterAtlasWrapper is Ownable, MergedSinglePriceFeedAdapterWit
+        (, uint256 atlasLatestBlockTimestamp) = getTimestampsFromLatestUpdate();
         uint256 atlasAnswer = getValueForDataFeed(dataFeedId);
-        if (atlasAnswer == 0) {
-            return int256(baseAnswer);
-        }
-        (uint256 baseLatestTimestamp,) = IAdapter(BASE_ADAPTER).getTimestampsFromLatestUpdate();
-        (uint256 atlasLatestTimestamp,) = getTimestampsFromLatestUpdate();
 
-        if (atlasLatestTimestamp > baseLatestTimestamp - BASE_FEED_DELAY) {
-            return int256(atlasAnswer);
+        uint80 roundId = IFeed(BASE_FEED).latestRound();
+
+        uint256 numIter;
+        for (;;) {
+            (uint256 baseValue,, uint128 baseBlockTimestamp) =
+                IAdapterWithRounds(BASE_ADAPTER).getRoundDataFromAdapter(dataFeedId, roundId);
+            if (atlasLatestBlockTimestamp >= baseBlockTimestamp) {
+                return int256(atlasAnswer);
+            }
+            if (block.timestamp - baseBlockTimestamp >= BASE_FEED_DELAY) {
+                return int256(baseValue);
+            }
+
+            unchecked {
+                numIter++;
+                roundId--;
+            }
+
+            if (numIter > MAX_HISTORICAL_FETCH_ITERATIONS) {
+                revert CannotFetchHistoricalData();
+            }
         }
-        return int256(baseAnswer);
+        // unreachable
+        return 0;
     }
 
     function latestTimestamp() public view virtual returns (uint256) {
-        (uint256 baseLatestTimestamp,) = IAdapter(BASE_ADAPTER).getTimestampsFromLatestUpdate();
-        (uint256 atlasLatestTimestamp,) = getTimestampsFromLatestUpdate();
-        uint256 atlasAnswer = getValueForDataFeed(getDataFeedId());
+        bytes32 dataFeedId = getDataFeedId();
+        (, uint256 atlasLatestBlockTimestamp) = getTimestampsFromLatestUpdate();
 
-        if (atlasAnswer == 0) {
-            return baseLatestTimestamp;
-        }
-        if (atlasLatestTimestamp > baseLatestTimestamp - BASE_FEED_DELAY) {
-            return atlasLatestTimestamp;
+        uint80 roundId = IFeed(BASE_FEED).latestRound();
+
+        uint256 numIter;
+        for (;;) {
+            (,, uint128 baseBlockTimestamp) =
+                IAdapterWithRounds(BASE_ADAPTER).getRoundDataFromAdapter(dataFeedId, roundId);
+            if (atlasLatestBlockTimestamp >= baseBlockTimestamp) {
+                return atlasLatestBlockTimestamp;
+            }
+            if (block.timestamp - baseBlockTimestamp >= BASE_FEED_DELAY) {
+                return uint256(baseBlockTimestamp);
+            }
+
+            unchecked {
+                numIter++;
+                roundId--;
+            }
+
+            if (numIter > MAX_HISTORICAL_FETCH_ITERATIONS) {
+                revert CannotFetchHistoricalData();
+            }
         }
-        return baseLatestTimestamp;
+        // unreachable
+        return 0;
     }
 
     function latestRoundData()
```

Comment:
I think this function needs the same historical data logic as implemented above in latestTimestamp() and latestAnswer().

Reply:
This calls `latestAnswerAndTimestamp`